crates/bevy_hierarchy/src/hierarchy.rs

```diff
@@ -20,6 +23,29 @@ pub struct DespawnChildrenRecursive {
     pub entity: Entity,
 }
 
+/// System to print a warning if at any time we detect an inconsistent hierarchy.
+pub fn hierarchy_healthcheck_system<T: Component>(
+    parent_query: Query<&Parent, (With<T>, Or<(Changed<Parent>, Added<T>)>)>,
+    component_query: Query<(), With<T>>,
+    mut already_diagnosed: Local<bool>,
+) {
+    if *already_diagnosed {
+        return;
+    }
+    for parent_of_component in &parent_query {
+        let parent = parent_of_component.get();
+        if !component_query.contains(parent) {
+            *already_diagnosed = true;
+            warn!(
+                "warning[B0004]: An entity with the {ty_name} component has a parent without {ty_name}.\n\
+                This will cause inconsistent behaviors! See https://bevyengine.org/learn/errors/#B0004",
+                ty_name=get_short_name(std::any::type_name::<T>()),
+            );
```

Comment:
How often do people use `Name` components on their entities? As this is for quicker debugging, it might be cool to optionally query the name and show it for the two entities if present.

Reply:
I like the idea.